### ISSUE TITLE
Queries not compatible with EF6

### ISSUE THREAD

**Issue by wonglae** [EntityGraphQL/EntityGraphQL#363]
As we verified below query doesn't work in EF 6 (.Net Core) but working on EF core.

Query = "{ people { foods { id } } }"

Field 'people' - LINQ to Entities does not recognize the method 'System.Collections.Generic.List`1[Dynamic_foods_c144bbbb-c568-462d-9908-1ed4b47af815] ToListWithNullCheck[Dynamic_foods_c144bbbb-c568-462d-9908-1ed4b47af815](System.Collections.Generic.IEnumerable`1[Dynamic_foods_c144bbbb-c568-462d-9908-1ed4b47af815], Boolean)' method, and this method cannot be translated into a store expression.

DB schema:

type Query {
	configuration: DbContextConfiguration
	"""Return a Food by its Id"""
	food(id: Int!): Food
	foods: [Food!]
	people: [Person!]
	"""Return a Person by its Id"""
	person(id: Int!): Person
}

type DbContextConfiguration {
	autoDetectChangesEnabled: Boolean!
	disableFilterOverProjectionSimplificationForCustomFunctions: Boolean!
	ensureTransactionsForFunctionsAndCommands: Boolean!
	lazyLoadingEnabled: Boolean!
	proxyCreationEnabled: Boolean!
	useDatabaseNullSemantics: Boolean!
	validateOnSaveEnabled: Boolean!
}

type Food {
	id: Int!
	name: String
	person: Person!
}

type Person {
	"""Return a Food by its Id"""
	food(id: Int!): Food
	foods: [Food!]
	id: Int!
	name: String
}

**Comment by lukemurray** [comment]
Sorry, can you clarify the versions? EF Core 6 is both version 6 and for .net core. Also .net core is no longer called .net core :)

If you mean EF from the older (non-cross platform) .net then that has not been supported. Support started with dotnet core 2 and has followed the release of that (dotnet core 2.1, dotnet core 3.1, dotnet 6, dotnet 7, dotnet 8, etc)

Can you post the specific dotnet version you are using and the specific EF reference?

Also what, if any, execution options you are passing to EntityGraphQL.

**Comment by wonglae** [comment]
What I meant is the older one (https://www.nuget.org/packages/EntityFramework). Microsoft eventually migrated ef6 to .Net Core (We consider .Net 5+ versions are all .Net Core versions :)) so it's cross platform now. Since this framework is claimed to not depend on EF Core we thought we can give a try using it. There are a lot of project like us still using legacy ef. Migrating to EF core will cost us years we probably will stay using legacy ef for a couple year. As I tried the older EntityGraphQL version 4.2.1 works fine. It sems to be a regression introduced by https://github.com/EntityGraphQL/EntityGraphQL/commit/6aa487d491221679b5f4a591a8ca8073ccfc6a65

**Comment by wonglae** [comment]
I was using default execution option. BTW we are not service app, we are a client app, thus we only consume the core part of this framework:

```
            var schemaProvider = SchemaBuilder.FromObject<AccountDbContext>();
            var query = new QueryRequest();
            query.Query = $"{{ sample(sort: [{{ id: ASC }}]) {{ items {{ id }} }} }}";
            var readResults = await schemaProvider.ExecuteRequestWithContextAsync(query, dbContext, null, null);
```

**Comment by lukemurray** [comment]
Seems like old EF doesn't like any extra methods. We basically build an expression like

`db.Sample.Sort(s => s.Id).Select(s => new { items = new { id = s.Items.Id } }).ToListWithNullCheck()`

EF Core is smart enough to ignore `ToListWithNullCheck()` and parse `db.Sample.Sort(s => s.Id).Select(s => new { items = new { id = s.Items.Id } })` into SQL etc. then call it on the result.

Seems old EF doesn't like that. I feel there might be other cases that old EF won't like as well.

**Comment by wonglae** [comment]
Yes, methods like ToListWithNullCheck are already precompiled by compiler and the old ef doesn't support. In our projects we define expressions like this:

```
                Expression<Func<MyEntity, bool>> expr = e =>
                    ((e.SyncState == ObjectState.New) ||
                    (e.SyncState == ObjectState.Deleted) ||
                    (e.SyncState == ObjectState.Updated) ||
                    e.MyAssociations.Any(
                        a => a.RowState == RowState.Committed &&
                            ((a.SyncState == ObjectState.New) ||
                            (a.SyncState == ObjectState.Deleted) ||
                            (a.SyncState == ObjectState.Updated)));
```

**Comment by lukemurray** [comment]
I haven't got to this, but I'd be open for a PR that fixes this.

**Comment by bzbetty** [comment]
thinking about ToListWithNullCheck. I'm actually pondering if it'd somehow be benefitial to be able to use ToListAsync - however that's an EFCore fature, so obviulsy not directly compatible with this library.

Wondering if the fix for both is to allow user to bring their own implementation?

**Comment by lukemurray** [comment]
Be keen to see if the latest changes in master help resolve your issue here. The change makes it only use `ToListWithNullCheck` when we know we should. Not sure how much it will help EF6 though.

**Comment by lukemurray** [comment]
Closing as changes have been made to better support this. Although EF6 (not Core) is not fully supported.
